Alarm for excessive number of requests blocked by WAF (#5286)

dsotirho-ucsc opened the pull request:
<!--
This is the PR template for regular PRs against `develop`. Edit the URL in your
browser's location bar, appending either `&template=promotion.md`,
`&template=hotfix.md`, `&template=backport.md` or `&template=gitlab.md` to
switch the template.
-->

Connected issues: #5286


## Checklist


### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title


### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] PR and connected issue are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>


### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>


### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>


### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>


### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>


### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>


### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.


### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer


### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.


### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator


### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Added commit title tags to merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub


### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>


### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.


### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing</sub>


### Operator

- [x] Unassigned PR


## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

dsotirho-ucsc committed:
[u] Alarm for excessive number of requests blocked by WAF (#5286)

dsotirho-ucsc commented:
<details>

<summary>Terraform plan</summary>

```
Terraform will perform the following actions:

  # aws_cloudwatch_metric_alarm.waf will be created
  + resource "aws_cloudwatch_metric_alarm" "waf" {
      + actions_enabled                       = true
      + alarm_actions                         = [
          + "arn:aws:sns:us-east-1:122796619775:azul-monitoring-dev",
        ]
      + alarm_name                            = "azul-waf-daniel"
      + arn                                   = (known after apply)
      + comparison_operator                   = "GreaterThanThreshold"
      + dimensions                            = {
          + "Region" = "us-east-1"
          + "Rule"   = "ALL"
          + "WebACL" = "azul-api_gateway-daniel"
        }
      + evaluate_low_sample_count_percentiles = (known after apply)
      + evaluation_periods                    = 1
      + id                                    = (known after apply)
      + metric_name                           = "BlockedRequests"
      + namespace                             = "AWS/WAFV2"
      + ok_actions                            = [
          + "arn:aws:sns:us-east-1:122796619775:azul-monitoring-dev",
        ]
      + period                                = 300
      + statistic                             = "Sum"
      + tags                                  = {
          + "Name"       = "azul-waf-daniel"
          + "billing"    = "hca"
          + "component"  = "azul-waf"
          + "deployment" = "daniel"
          + "owner"      = "dsotirho@ucsc.edu"
          + "service"    = "azul"
        }
      + tags_all                              = {
          + "Name"       = "azul-waf-daniel"
          + "billing"    = "hca"
          + "component"  = "azul-waf"
          + "deployment" = "daniel"
          + "owner"      = "dsotirho@ucsc.edu"
          + "service"    = "azul"
        }
      + threshold                             = 1500
      + treat_missing_data                    = "notBreaching"
    }
```

</details>


## Blocked requests (sum over 5 minutes)

### Personal deployments:
![waf-personal](https://github.com/DataBiosphere/azul/assets/46407464/46071846-341f-4003-bb61-36e3b8420b00)

### dev:
![waf-dev](https://github.com/DataBiosphere/azul/assets/46407464/4772f6ab-223f-4d49-80b4-ef1e0bc914ee)

### anvildev:
![waf-anvildev](https://github.com/DataBiosphere/azul/assets/46407464/71036083-3dcf-4423-bf5c-a17637204585)

### sandbox:
![waf-sandbox](https://github.com/DataBiosphere/azul/assets/46407464/097b8c91-c6f7-464e-9076-cb2111286131)

### anvilbox:
![waf-anvilbox](https://github.com/DataBiosphere/azul/assets/46407464/3db72ba6-5395-41d7-8172-d03bb7d11caa)

### hammerbox:
![waf-hammerbox](https://github.com/DataBiosphere/azul/assets/46407464/f4d0d4bc-1fd3-432f-9566-50ee9fc656b7)

### prod:
(Image omitted for security purposes)
Max value: 5,192
Common spike: 1,000 - 3,000

### anvilprod:
(Image omitted for security purposes)
Max value: 8,928
Common spike: 500 - 1,000

achave11-ucsc reviewed (commented):
Changes look good. 
However, I'd like for this changes to be tested in `anvildev` or `dev`. We should trigger a reindex, (it be nice to run the locus script also, to simulate, say 2 users and see some service traffic) to ensure we won't be bugged down with more alarms of this sort, of obscure provenance.

Also, does it seem strange all those high WAF blocked request for the lower deployments?

dsotirho-ucsc commented:
The WAF blocked request counts have increased a bit in the last two weeks (see `anvildev` & `anvilbox`). I've updated the screenshots in [my previous post](https://github.com/DataBiosphere/azul/pull/5455#issuecomment-1670458279) and adjusted the thresholds set in this PR accordingly.

achave11-ucsc reviewed (changes requested):
Changes look good, however upgrading instructions need to be carried out by developers and the PR CL is missing an item to remind everyone.

dsotirho-ucsc commented:
> the PR CL is missing an item to remind everyone

Added a reminder CL item immediately after `Pushed merge commit to GitHub`

hannes-ucsc commented on environment.py:
```diff
@@ -733,5 +733,12 @@ def env() -> Mapping[str, Optional[str]]:
+        # can be found by looking at the WAF specific aws_cloudwatch_metric_alarm
+        # resource in the Terraform config files.
+        #
+        'AZUL_MAX_WAF_BLOCKED_REQUESTS': '0',
```
Tying the threshold to a deployment is coincidental. Which deployment becomes victim to an attack is entirely out of our hands. It could hit production or a personal deployment. Ideally we'd make the metric a ratio between the numbers of served and blocked requests. If that's not possible we need to use a threshold that's "normal" for `prod`. In that case, please request a PL slot to discuss what that "norm" is.

dsotirho-ucsc commented:
Note, checking this metric query on `prod` shows that this alarm would of fired a few times in the last 4 weeks.

![Screen Shot 2023-09-14 at 2 05 37 AM](https://github.com/DataBiosphere/azul/assets/46407464/19150b6b-297d-4730-b948-5d67d066c8a1)

hannes-ucsc commented on environment.py:
```diff
@@ -733,5 +733,5 @@ def env() -> Mapping[str, Optional[str]]:
         # })
         #
-        'azul_slack_integration': None
+        'azul_slack_integration': None,
```
Remove.

hannes-ucsc commented on terraform/cloudwatch.tf.json.template.py:
```diff
@@ -313,6 +313,41 @@ def prod_qualified_resource_name(name: str) -> str:
+                        'waf_blocked': {
+                            'alarm_name': config.qualified_resource_name('waf_blocked'),
+                            'comparison_operator': 'GreaterThanThreshold',
+                            'threshold': 50,
```
Does this mean we'd be allowing 50% of the request to be blocked?

dsotirho-ucsc commented:
> Does this mean we'd be allowing 50% of the request to be blocked?

Yes, the current settings would trigger the alarm when greater than 50% of requests are blocked, however it looks like this might cause the alarm to trigger quite frequently.

[Metric query % blocked on hcadev](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100~label~'Expression1~id~'e1))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-dev~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~300~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d)

![blocked-hca-dev](https://github.com/DataBiosphere/azul/assets/46407464/3108ec61-5072-4a6f-8b88-1fafd6fcfae3)

[Metric query % blocked on hcaprod](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100~label~'Expression1~id~'e1~stat~'Sum))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-prod~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~300~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d)

![blocked-hca-prod](https://github.com/DataBiosphere/azul/assets/46407464/4244dc93-1f1c-48e6-b4e5-3a801d4b010c)

[Metric query % blocked on anvildev](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100~label~'Expression1~id~'e1))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-anvildev~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~300~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d)

![blocked-anvil-dev](https://github.com/DataBiosphere/azul/assets/46407464/b54cdd06-c045-44ae-892b-0b7cd16f4bf3)

[Metric query % blocked on anvilprod](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100~label~'Expression1~id~'e1))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-anvilprod~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~300~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d)

![blocked-anvil-prod](https://github.com/DataBiosphere/azul/assets/46407464/52e044db-b920-49be-91e7-e7b2409bdb23)

hannes-ucsc reviewed (changes requested):
>> Does this mean we'd be allowing 50% of the request to be blocked?

> Yes, the current settings would trigger the alarm when greater than 50% of requests are blocked, however it looks like this might cause the alarm to trigger quite frequently.

That wouldn't work, obviously. What's your plan on addressing this?

dsotirho-ucsc commented:
Decided in parking lot discussion the alarm should trip when a threshold of 25% requests blocked is exceeded for 4 out of 4 evaluation periods, each period spanning 15 minutes.

With this metric query, the alarm would not have tripped in the last four weeks, however it would have come close with 3 consecutive 15 minute periods exceeding 25% blocked requests occurring twice on `hcaprod` (09/10 and 09/11) within the last 4 weeks, and once on `anvildev` (09/03).

[% blocked requests > 25% on hcadev](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100*20*3e*2025~label~'Expression1~id~'e1~period~900))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-dev~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~900~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d):

![blocked-hcadev](https://github.com/DataBiosphere/azul/assets/46407464/ffa932be-b87d-42fe-b451-1ce1e97624ea)

[% blocked requests > 25% on hcaprod](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100*20*3e*2025~label~'Expression1~id~'e1~stat~'Sum~period~900))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-prod~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~900~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d):

![blocked-hcaprod](https://github.com/DataBiosphere/azul/assets/46407464/ec04f4a3-713b-4159-b1c3-84162c5523cf)

[% blocked requests > 25% on anvildev](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100*20*3e*2025~label~'Expression1~id~'e1~period~900))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-anvildev~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~900~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d):

![blocked-anvildev](https://github.com/DataBiosphere/azul/assets/46407464/9cfd0d96-51cb-4b4d-9d11-6fcc0cc703ea)

[% blocked requests > 25% on anvilprod](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100*20*3e*2025~label~'Expression1~id~'e1~period~900))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-anvilprod~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~900~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d):

![blocked-anvilprod](https://github.com/DataBiosphere/azul/assets/46407464/f820c804-1017-4a3f-981a-60e75ac41aa8)

dsotirho-ucsc commented:
Creation of the alarm tested on my personal deployment.

![Screen Shot 2023-09-20 at 6 12 52 PM](https://github.com/DataBiosphere/azul/assets/46407464/74c0b7b2-37f5-4914-a835-fb57ad362b26)
